### PR TITLE
Perform token-substitution in all String-attributes of a check

### PR DESCRIPTION
**Pull request by UnitedMarsupials**
See #1281 . Thanks!

### CODE REVIEW

**Comment by UnitedMarsupials** [comment]
Sorry, seems like my editor inserted a tab in one line instead of 8 spaces as seems to be Sensu's coding style. The comment should, probably, also be edited to say "a**n** object".

But I don't want to commit these decorative changes myself to avoid jeopardizing this feature making it into 0.24 as promised by @portertech in #1281 :-) Thanks!

File: lib/sensu/client/process.rb [hunk 1]
```diff
@@ -99,6 +99,37 @@ def publish_check_result(check)
...
+      # Perform token-substitution for a object. Strings
+      # are passed to substitute_tokens(), arrays and sub-hashes are
+      # processed recursively. Numbers (and other types?) are skipped
+      # at the moment, because substitute_tokens() can not handle them
+      # anyway.
+      #
+      # @param	[Object]
+      # @return	[Object] updated object of the same type as the argument
+      # 	plus accumulated list of unmatched tokens
+      def obj_substitute_tokens(obj)
```
Comment (portertech) [inline lib/sensu/client/process.rb]: Let's drop `obj_` from the method name :+1:

File: lib/sensu/client/process.rb [hunk 2]
```diff
@@ -99,6 +99,37 @@ def publish_check_result(check)
...
+          obj.map! {|value|
+            value, unmatched = obj_substitute_tokens(value)
+            unmatched_tokens.push(*unmatched)
+            value
+	  }
+        when String
+          obj, unmatched_tokens = substitute_tokens(obj, @settings[:client])
+        end
+	unmatched_tokens.uniq! if unmatched_tokens
+        [obj, unmatched_tokens]
```
Comment (portertech) [inline lib/sensu/client/process.rb]: Just do `unmatched_tokens.uniq` here?

File: lib/sensu/client/process.rb (hunk 1 above)
Comment (portertech) [inline lib/sensu/client/process.rb]: Maybe `obj` -> `object` or `definition`?

**Comment by portertech** [comment]
@UnitedMarsupials need the PR to target the release/0.24 branch.

File: lib/sensu/client/process.rb [hunk 3]
```diff
@@ -99,6 +99,37 @@ def publish_check_result(check)
...
+        case obj
+        when Hash
+          unmatched_tokens = []
+          obj.each do |key, value|
+            obj[key], unmatched = obj_substitute_tokens(value)
+            unmatched_tokens.push(*unmatched)
+          end
+        when Array
+          unmatched_tokens = []
+          obj.map! {|value|
```
Comment (portertech) [inline lib/sensu/client/process.rb]: Keep to the `do` `end` style that  you used for Hash?

**Comment by UnitedMarsupials** [comment]
> Let's drop obj_ from the method name

But then it will conflict with the original `substitute_tokens` from `utilities.rb`, no?

> Just do unmatched_tokens.uniq here?

Where?

> need the PR to target the release/0.24 branch.

Sorry, I don't understand... Oh, you mean, pull-request must be against 0.24, rather than master branch? Ok...

**Comment by UnitedMarsupials** [comment]
How about now? The `obj_` prefix is still there -- open to suggestions...

**Comment by UnitedMarsupials** [comment]
Maybe, this new method of mine should replace the old, string-only `substitute_tokens`? Incorporating the old code inside its `when String` branch. But then it will also apply to the filter-handling -- which may or may not be desirable...

**Comment by portertech** [comment]
Merged https://github.com/sensu/sensu/pull/1294 :+1:

**Pull request closed without merging**